src/main/java/org/klodnicki/repository/BookRepository.java

```diff
@@ -1,16 +1,68 @@
 package org.klodnicki.repository;
 
 import jakarta.persistence.EntityManager;
-import org.klodnicki.entity.Book;
+import jakarta.persistence.NoResultException;
+import jakarta.persistence.NonUniqueResultException;
+import jakarta.persistence.TypedQuery;
+import org.klodnicki.entity.BookInfo;
+
+import java.util.List;
+import java.util.Optional;
 
 public class BookRepository {
 
     private final EntityManager entityManager = HibernateUtil.getSessionFactory().createEntityManager();
 
-    public void add(Book book) {
+    public void add(BookInfo bookInfo) {
         entityManager.getTransaction().begin();
-        entityManager.persist(book);
+        entityManager.persist(bookInfo);
         entityManager.getTransaction().commit();
     }
 
+    public Optional<BookInfo> findBookByTitleAndAuthor(String title, String author) {
+        String hqlQuery = "FROM BookInfo b WHERE b.title = :title AND b.author = :author";
+        TypedQuery<BookInfo> query = entityManager.createQuery(hqlQuery, BookInfo.class);
+        query.setParameter("title", title);
+        query.setParameter("author", author);
+        // 1 solution: always return List of objects
+
+        // 2 solution: returning optional. TO BE COMPLETED with orElseThrow
+        BookInfo singleResult = null;
+        try {
+            singleResult = query.getSingleResult();
+        } catch (NoResultException e) {
+            return Optional.empty();
+        } catch (NonUniqueResultException e) {
+            // Returns first element
+            for (BookInfo bookInfo : query.getResultList()) {
+                return Optional.of(bookInfo);
+            }
+        }
+        return Optional.ofNullable(singleResult);
+    }
+
+    public BookInfo findBookByTitleAndAuthorAndEdition(String title, String author, String edition) {
```

Comment:
This method also shouldn't just return `BookInfo`. Same reason as `findBookByTitleAndAuthor`. If you want to catch exceptions thrown by `getSingleResult` do it here.